crates/precompile/src/blake2.rs

```diff
@@ -113,21 +162,540 @@ pub mod algo {
             v[14] = !v[14] // Invert all bits if the last-block-flag is set.
         }
         for i in 0..rounds {
-            // Message word selection permutation for this round.
-            let s = &SIGMA[i % 10];
-            g(&mut v, 0, 4, 8, 12, m[s[0]], m[s[1]]);
-            g(&mut v, 1, 5, 9, 13, m[s[2]], m[s[3]]);
-            g(&mut v, 2, 6, 10, 14, m[s[4]], m[s[5]]);
-            g(&mut v, 3, 7, 11, 15, m[s[6]], m[s[7]]);
-
-            g(&mut v, 0, 5, 10, 15, m[s[8]], m[s[9]]);
-            g(&mut v, 1, 6, 11, 12, m[s[10]], m[s[11]]);
-            g(&mut v, 2, 7, 8, 13, m[s[12]], m[s[13]]);
-            g(&mut v, 3, 4, 9, 14, m[s[14]], m[s[15]]);
+            round(&mut v, &m, i);
         }
 
         for i in 0..8 {
             h[i] ^= v[i] ^ v[i + 8];
         }
     }
+
+    #[inline(always)]
+    fn round(v: &mut [u64; 16], m: &[u64; 16], r: usize) {
+        // Message word selection permutation for this round.
+        let s = &SIGMA[r % 10];
+        // g1
+        g(v, 0, 4, 8, 12, m[s[0]], m[s[1]]);
+        g(v, 1, 5, 9, 13, m[s[2]], m[s[3]]);
+        g(v, 2, 6, 10, 14, m[s[4]], m[s[5]]);
+        g(v, 3, 7, 11, 15, m[s[6]], m[s[7]]);
+
+        // g2
+        g(v, 0, 5, 10, 15, m[s[8]], m[s[9]]);
+        g(v, 1, 6, 11, 12, m[s[10]], m[s[11]]);
+        g(v, 2, 7, 8, 13, m[s[12]], m[s[13]]);
+        g(v, 3, 4, 9, 14, m[s[14]], m[s[15]]);
+    }
+}
+
+// Adapted from https://github.com/rust-lang-nursery/stdsimd/pull/479.
+macro_rules! _MM_SHUFFLE {
+    ($z:expr, $y:expr, $x:expr, $w:expr) => {
+        ($z << 6) | ($y << 4) | ($x << 2) | $w
+    };
+}
+
+/// Code adapted from https://github.com/oconnor663/blake2_simd/blob/82b3e2aee4d2384aabbeb146058301ff0dbd453f/blake2b/src/avx2.rs
+#[cfg(target_feature = "avx2")]
```

Comment:
you can use std::is_x86_feature_detected for runtime detection instead of only compile time

Reply:
the macro resolves true at compile time if -Ctarget-cpu / -Ctarget-feature is passed, otherwise it resolves at run time

Reply:
This is cool! Including it will mean we don't need any additional input on compilation, and the switch is done internally at runtime.

Reply:
you're missing `#[target_feature(enable = "avx2")]` on the function. if you do both cfg() and is_detected then the is_detected has no impact